extensions/_modules/heroku.py

```diff
@@ -0,0 +1,214 @@
+# -*- coding: utf-8 -*-
+'''
+Module for managing heroku apps.
+
+.. versionadded:: 2015.5.0
+
+:configuration: This module can be used by either passing an api key and version
+    directly or by specifying both in a configuration profile in the salt
+    master/minion config.
+
+    For example:
+
+    .. code-block:: yaml
+
+        heroku:
+          api_key: peWcBiMOS9HrZG15peWcBiMOS9HrZG15
+
+    Custom API Example:
+
+    .. code-block:: yaml
+
+        heroku:
+          api_url: http://api.heroku.myteam.com
+          api_key: peWcBiMOS9HrZG15peWcBiMOS9HrZG15
+'''
+# Import Python Libs
+from __future__ import absolute_import
+import json
+import logging
+
+# Import 3rd-party Libs
+# pylint: disable=import-error,no-name-in-module,redefined-builtin
+from salt.ext.six.moves.urllib.parse import urljoin as _urljoin
+from salt.ext.six.moves.urllib.parse import urlencode as _urlencode
+from salt.ext.six.moves import range
+import salt.ext.six.moves.http_client
+
+import salt.utils.http
+
+# pylint: enable=import-error,no-name-in-module,redefined-builtin
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'heroku'
+
+def __virtual__():
+    '''
+    Return virtual name of the module.
+
+    :return: The virtual name of the module.
+    '''
+    return __virtualname__
+
+def _query(
+       api_url=None,
+       api_key=None,
+       endpoint=None,
+       arguments=None,
+       method='GET',
+       data=None):
+  '''
+  Heroku object method function to construct and execute on the API URL.
+
+  :param api_url:     The Heroku API URL.
+  :param api_key:     The Heroku api key.
+  :param function:    The Heroku api function to perform.
+  :param method:      The HTTP method, e.g. GET, POST, or PATCH.
+  :param data:        The data to be sent for POST/PATCH method.
+  :return:            The json response from the API call or False.
+  '''
+  headers = {}
+  query_params = {}
+
+  if not api_url:
+      try:
+          options = __salt__['config.option']('heroku')
+          api_url = options.get('api_url')
+      except (NameError, KeyError, AttributeError):
+          pass  # not mandatory, thus won't fail if not found
+
+  if not api_key or not endpoint:
+      try:
+          options = __salt__['config.option']('heroku')
+          if not api_key:
+            api_key = options.get('api_key')
+          if not endpoint:
+            endpoint = options.get('endpoint')
+      except (NameError, KeyError, AttributeError):
+          log.error("No Heroku api key or endpoint found.")
+          return False
+
+  use_api_url = 'https://api.heroku.com'  # default API URL
+  if api_url:
+      use_api_url = api_url
```

Comment:
It would be cleaner to get rid of this section and just set the default value for `api_url` in the function signature.